runtime-config.md

```diff
@@ -10,7 +10,7 @@ Presently there are `Prestart`, `Poststart` and `Poststop`.
 * [`Poststop`](#poststop) is a list of hooks to be run after the container process exits
 
 Hooks allow one to run code before/after various lifecycle events of the container.
-Hooks MUST be called in the listed order.
+Hooks MUST be called in the listed order, with each hook process being reaped before the next hook is executed.
```

Comment:
On Thu, Dec 10, 2015 at 10:55:18PM -0800, Qiang Huang wrote:
> > -Hooks MUST be called in the listed order.
> > +Hooks MUST be called in the listed order, with each hook process being reaped before the next hook is executed.
>
> Wouldn't this cause confusion if "each hook process" includes the
> child processes it creates?

What sort of confusion?  If the hook wants to explicitly launch a
daemon or some such, I'm fine with:

  hook 1: spawn ------> reaped
                daemon------------------------------------------->
  hook 2:                      spawn --> reaped
  hook 3:                                       spawn --> reaped

If folks are worried about hook-children surviving an early hook exit:

  hook 1: spawn ------(died)-> reaped
                child------------------------------------------->
  hook 2:                             spawn --> reaped
  hook 3:                                              spawn --> reaped

we can require the runtime to be more shell-like and make a new
process group to contain the hooks (but this will start getting pretty
complicated).

Without some sort of rule about reaping, I don't see much gain to the
“Hooks MUST be called in the listed order” requirement.  If they're
all spawned one after the other, you don't have any guarantees that
earlier hooks will have accomplished their task before later hooks are
run.

It might be easier to just separate ‘create’ (most of the container
setup) from ‘exec’ (process launch and possible PID namespace
creation) and ‘destroy’ (remove anything left-over from ‘create’) and
get out of the hook business altogether [1,2,3].

[1]: https://github.com/opencontainers/specs/issues/115#issuecomment-147553272
[2]: https://github.com/opencontainers/specs/issues/226
[3]: http://ircbot.wl.linuxfoundation.org/meetings/opencontainers/2015/opencontainers.2015-10-28-17.02.log.html#l-71